ocaml/lib/snarky_js_bindings_lib.ml

```diff
@@ -1248,563 +1219,163 @@ let poseidon =
-                new%js field_constr (As_field.of_field (conv x)) ) ) )
-    in
-    Typ.transport
-      (Typ.array ~length:typ##sizeInFields Field.typ)
-      ~there:(to_array (fun x -> Option.value_exn (Field.to_constant x)))
-      ~back:(of_array Field.constant)
-    |> Typ.transport_var ~there:(to_array Fn.id) ~back:(of_array Fn.id)
-
-  let witness (type a) (typ : a as_field_elements Js.t)
-      (f : (unit -> a) Js.callback) : a =
-    let a =
-      Impl.exists (typ_ typ) ~compute:(fun () : a -> Js.Unsafe.fun_call f [||])
-    in
-    if Js.Optdef.test (Js.Unsafe.coerce typ)##.check then
-      let () = (Js.Unsafe.coerce typ)##check a in
-      ()
-    else failwith "Circuit.witness: input does not have a `check` method" ;
-    a
-
-  let typ_minimal (type a) (typ : a as_field_elements_minimal Js.t) =
-    Typ.array ~length:typ##sizeInFields Field.typ
-
-  let witness_minimal (type a) (typ : a as_field_elements_minimal Js.t)
-      (f : (unit -> field_class Js.t Js.js_array Js.t) Js.callback) =
-    Impl.exists (typ_minimal typ) ~compute:(fun () ->
-        Js.Unsafe.fun_call f [||]
-        |> Js.to_array
-        |> Array.map ~f:of_js_field_unchecked )
+module Snarky = struct
+  let typ (size_in_fields : int) = Typ.array ~length:size_in_fields Field.typ
+
+  let exists (size_in_fields : int)
+      (compute : unit -> field_class Js.t Js.js_array Js.t) =
+    Impl.exists (typ size_in_fields) ~compute:(fun () ->
+        compute () |> Js.to_array |> Array.map ~f:of_js_field_unchecked )
     |> Array.map ~f:to_js_field |> Js.array
 
-  module Circuit_main = struct
-    type ('w, 'p) t =
-      < snarkyMain : ('w -> 'p -> unit) Js.callback Js.prop
-      ; snarkyWitnessTyp : 'w as_field_elements Js.t Js.prop
-      ; snarkyPublicTyp : 'p as_field_elements Js.t Js.prop >
-      Js.t
-  end
+  let as_prover = Impl.as_prover
 
-  let main_and_input (type w p) (c : (w, p) Circuit_main.t) =
-    let main ?(w : w option) (public : p) () =
-      let w : w =
-        witness c##.snarkyWitnessTyp
-          (Js.wrap_callback (fun () -> Option.value_exn w))
-      in
-      Js.Unsafe.(fun_call c##.snarkyMain [| inject w; inject public |])
-    in
-    (main, typ_ c##.snarkyPublicTyp)
+  let run_and_check (f : unit -> unit) =
+    try
+      Impl.run_and_check_exn (fun () ->
+          f () ;
+          fun () -> () )
+    with exn -> raise_exn exn
 
-  let generate_keypair (type w p) (c : (w, p) Circuit_main.t) :
-      keypair_class Js.t =
-    let main, input_typ = main_and_input c in
-    let cs =
-      Impl.constraint_system ~input_typ ~return_typ:Snark_params.Tick.Typ.unit
-        (fun x -> main x)
-    in
-    let kp = Impl.Keypair.generate ~prev_challenges:0 cs in
-    new%js keypair_constr kp
+  let run_unchecked (f : unit -> unit) =
+    try
+      Impl.run_and_check_exn (fun () ->
+          Snarky_backendless.Snark0.set_eval_constraints false ;
+          f () ;
+          Snarky_backendless.Snark0.set_eval_constraints true ;
```

Comment:
is it because of the global state that we have to switch between these modes? :D

Reply:
yes, in this function we want checks like `x.assertEquals(y)` to not throw (because we want to be able to use dummy data), and therefore we have to toggle this global flag